src/coreclr/jit/morph.cpp

```diff
@@ -13978,6 +13576,272 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
+
+            if (reverse)
+            {
+                gtReverseCond(op1);
+            }
+
+            noway_assert((op1->gtFlags & GTF_RELOP_JMP_USED) == 0);
+            op1->gtFlags |= cmp->gtFlags & (GTF_RELOP_JMP_USED | GTF_RELOP_QMARK | GTF_DONT_CSE);
+            op1->SetVNsFromNode(cmp);
+
+            DEBUG_DESTROY_NODE(cmp);
+            return op1;
+        }
+
+        //
+        // Now we check for a compare with the result of an '&' operator
+        //
+        // Here we look for the following transformation:
+        //
+        //                        EQ/NE                  EQ/NE
+        //                        /  \                   /  \.
+        //                      AND   CNS 0/1  ->      AND   CNS 0
+        //                     /   \                  /   \.
+        //                RSZ/RSH   CNS 1            x     CNS (1 << y)
+        //                  /  \.
+        //                 x   CNS_INT +y
+
+        if (fgGlobalMorph && op1->OperIs(GT_AND) && op1->AsOp()->gtGetOp1()->OperIs(GT_RSZ, GT_RSH))
+        {
+            GenTreeOp* andOp    = op1->AsOp();
+            GenTreeOp* rshiftOp = andOp->gtGetOp1()->AsOp();
+
+            if (!rshiftOp->gtGetOp2()->IsCnsIntOrI())
+            {
+                goto SKIP;
+            }
+
+            ssize_t shiftAmount = rshiftOp->gtGetOp2()->AsIntCon()->IconValue();
+
+            if (shiftAmount < 0)
+            {
+                goto SKIP;
+            }
+
+            if (!andOp->gtGetOp2()->IsIntegralConst(1))
+            {
+                goto SKIP;
+            }
+
+            GenTreeIntConCommon* andMask = andOp->gtGetOp2()->AsIntConCommon();
+
+            if (andOp->TypeIs(TYP_INT))
+            {
+                if (shiftAmount > 31)
+                {
+                    goto SKIP;
+                }
+
+                andMask->SetIconValue(static_cast<int32_t>(1 << shiftAmount));
+
+                // Reverse the condition if necessary.
+                if (op2Value == 1)
+                {
+                    gtReverseCond(cmp);
+                    op2->SetIconValue(0);
+                }
+            }
+            else if (andOp->TypeIs(TYP_LONG))
+            {
+                if (shiftAmount > 63)
+                {
+                    goto SKIP;
+                }
+
+                andMask->SetLngValue(1ll << shiftAmount);
+
+                // Reverse the cond if necessary
+                if (op2Value == 1)
+                {
+                    gtReverseCond(cmp);
+                    op2->SetLngValue(0);
+                }
+            }
+
+            andOp->gtOp1 = rshiftOp->gtGetOp1();
+
+            DEBUG_DESTROY_NODE(rshiftOp->gtGetOp2());
+            DEBUG_DESTROY_NODE(rshiftOp);
+        }
+    }
+
+SKIP:
+
+    // Now check for compares with small constant longs that can be cast to int.
+    if (!op2->TypeIs(TYP_LONG) || ((op2->LngValue() >> 31) != 0))
+    {
+        return cmp;
+    }
```

Comment:
I know this is preexisting code, but do you know why we filter out negative values even if they are in int range here? Seems at odds with the comment.

Reply:
The purpose of this is to only allow values for which sign extensions is the same as zero extension. This is required for the transform below (where we throw away the cast) to be correct, otherwise the comparisons in different widths would not agree.

Example: `EQ(-1L, CAST_UN(int))` is always `false`, but if we were to make it into `EQ(-1, int)`, it could now become `true` for negative inputs.

As a CQ opportunity, we could use the same logic we use for `MUL_LONG` recognition ([which makes it more explicit what is going on](https://github.com/dotnet/runtime/blob/d367827039f4120129317a97e4be413e29f57119/src/coreclr/jit/morph.cpp#L14742-L14748)).

Reply:
Maybe clarify this in the comment above?
For CQ we could also handle sign/zero extending cast instead of giving up on the transformation, I suppose. Not sure how valuable this would be, but perhaps something to consider in the future.

Reply:
> Maybe clarify this in the comment above?

Will do.

> For CQ we could also handle sign/zero extending cast instead of giving up on the transformation, I suppose.

Yes... we could also enable it for more trees in general, if we wanted. I have #55186 which will add an abstraction that helps with these scenarios.